segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/AsyncTableKeyBuilder.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.tables;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.TimeoutTimer;
+import io.pravega.common.io.EnhancedByteArrayOutputStream;
+import io.pravega.common.io.SerializationException;
+import io.pravega.common.io.StreamHelpers;
+import io.pravega.common.util.ArrayView;
+import io.pravega.segmentstore.contracts.ReadResultEntry;
+import io.pravega.segmentstore.contracts.ReadResultEntryContents;
+import io.pravega.segmentstore.contracts.ReadResultEntryType;
+import io.pravega.segmentstore.server.reading.AsyncReadResultHandler;
+import io.pravega.segmentstore.server.reading.AsyncReadResultProcessor;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import lombok.Getter;
+import lombok.NonNull;
+
+/**
+ * Helps builds Table Entry Keys asynchronously by processing successive {@link ReadResultEntry}. This can be used to
+ * read serialized Keys from a Segment starting at the Entry serialization offset but without knowing the length of the
+ * Key.
+ *
+ * NOTE: this class is meant to be used as an {@link AsyncReadResultHandler} by an {@link AsyncReadResultProcessor}, which
+ * processes it in a specific sequence using its own synchronization primitives. As such this class should be considered
```

Comment:
Yes. This is what this doc explains. 

If `AsyncReadResultProcessor` changes behavior, then anything which implements `AsyncReadResultHandler` will also need to change. This class is one of them. It would be really difficult to make this class work both with the current behavior and with a "future behavior" (presumably where the processor doesn't wait for completion before invoking next).